File: java-algorithm/src/main/java/com/leetcode/Problem926.java
package com.leetcode;

/**
 * 926. Flip String to Monotone Increasing
 * https://leetcode.com/problems/flip-string-to-monotone-increasing/
 */
public class Problem926 {
    public int minFlipsMonoIncr(String s) {
        int n = s.length();
        int[][] dp = new int[n][2];
        if (s.charAt(n - 1) == '0') {
            dp[n - 1][1] = 1;
        } else {
            dp[n - 1][0] = 1;
        }
        for (int i = n - 2; i >= 0; i--) {
            if (s.charAt(i) == '0') {
                dp[i][0] = Math.min(dp[i + 1][0], dp[i + 1][1]);
                dp[i][1] = dp[i + 1][1] + 1;
            } else {
                dp[i][0] = Math.min(dp[i + 1][0], dp[i + 1][1]) + 1;
                dp[i][1] = dp[i + 1][1];
            }
        }
        return Math.min(dp[0][0], dp[0][1]);
    }
}

File: python-algo/leetcode/problem_926.py
"""926. Flip String to Monotone Increasing
https://leetcode.com/problems/flip-string-to-monotone-increasing/
"""
import functools


class Solution:
    def minFlipsMonoIncr(self, s: str) -> int:
        @functools.lru_cache(None)
        def dfs(i: int, pre: str) -> int:
            if i == len(s):
                return 0
            if pre == '0' and s[i] == '0':
                ret = dfs(i + 1, s[i])
            elif pre == '0' and s[i] == '1':
                ret = min(dfs(i + 1, '1'), dfs(i + 1, '0') + 1)
            elif pre == '1' and s[i] == '0':
                ret = dfs(i + 1, '1') + 1
            else:
                ret = dfs(i + 1, '1')
            return ret

        return dfs(0, '0')
